src/analyzer/protocol/ssl/spicy/SSL.spicy

```diff
@@ -0,0 +1,2058 @@
+    gmt_unix_time: uint32 &parse-from=self.random_bytes;
+    session_id_length: uint8;
+    session_id: bytes &size=self.session_id_length;
+    cipher_suite: uint16;
+    compression_method: uint8;
+    extensions_length: uint16 if(len > self.offset() + 2);
+    extensions: Extension(sh, False)[] &size=self.extensions_length if(len > self.offset() + 2);
+
+    on cipher_suite {
+        sh.chosen_cipher = self.cipher_suite;
+        # print "set chosen cipher", self.cipher_suite, sh.chosen_cipher;
+    }
+};
+
+type Extension = unit(inout sh: Share, client_hello: bool) {
+    code: uint16;
+    length: uint16;
+    # position is re-set, just after parsing raw, so that the remainder of the unit can parse the raw data.
+    # raw data is needed for an event.
+    raw: bytes &size=self.length;
+
+    switch (Extensions(self.code)) {
+        # Extensions::next_protocol_negotiation -> :bytes &size=self.extension_size; # alert if != 0
+        Extensions::ec_point_formats -> ec_point_formats: EcPointsFormat_extension(sh) &max-size=self.length if(self.length > 0);
+        Extensions::supported_groups -> elliptic_curves: EllipticCurveList(sh) &max-size=self.length if(self.length > 0);
+        Extensions::SessionTicket_TLS -> ticket_data: bytes &size=self.length; # ticket data
+        Extensions::heartbeat -> heartbeat: uint8 &convert=HeartbeatMode($$) &max-size=self.length if(self.length > 0);
+        Extensions::signature_algorithms -> signature_algorithms: SignatureAlgorithms(sh) &max-size=self.length if(self.length > 0);
+        Extensions::renegotiation_info -> renegotiation_info: RenegotiationInfo &max-size=self.length if(self.length > 0);
+        Extensions::server_name -> server_name: ServerNameList(sh) &max-size=self.length if(self.length > 0);
+        Extensions::application_layer_protocol_negotiation -> application_layer_protocol_negotiation: ProtocolNameList(sh) &max-size=self.length if(self.length > 0);
+        Extensions::status_request -> status_request: StatusRequest(self.length) &max-size=self.length if(self.length > 0);
+        Extensions::signed_certificate_timestamp -> signed_certificate_timestamp: SignedCertificateTimestampList(sh) &max-size=self.length if(self.length > 0);
+        Extensions::key_share_old -> key_share_old: KeyShare(client_hello, sh, self.length) &max-size=self.length if(self.length > 0);
+        Extensions::key_share -> key_share: KeyShare(client_hello, sh, self.length) &max-size=self.length if(self.length > 0);
+        Extensions::supported_versions -> supported_versions: SupportedVersionsSelector(sh, client_hello) &max-size=self.length if(self.length > 0);
+        Extensions::psk_key_exchange_modes -> psk_key_exchange_modes: PSKKeyExchangeModes(sh) &max-size=self.length if(self.length > 0);
+        Extensions::pre_shared_key -> pre_shared_key: PreSharedKey(sh, client_hello) &max-size=self.length if(self.length > 0);
+        * -> unknown: bytes &size=self.length;
+    };
```

Comment:
Are those `&max-size` and `&if` all same? Think they could work on the outer switch, too? Saw this somewhere else, too.

Also wonder if you could go with `&parse-from=self.raw` instead of the `set_input()`, so that wouldn't require random access. Not sure that makes a huge difference though.

Reply:
Sounds nice, but doesn't work:

```
[error] /Users/johanna/zeek/main/src/analyzer/protocol/ssl/spicy/SSL.spicy:1109:5-1127:28: attribute '&max-size' is not supported here
```